pkg/tfbridge/detailed_diff_test.go

```diff
@@ -0,0 +1,1631 @@
+		},
+		{
+			name: "tf force new unchanged",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("old"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: nil,
+		},
+		{
+			name: "tf force new changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE},
+		},
+		{
+			name: "tf force new changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE},
+		},
+		{
+			name: "tf force new changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			etf:  shimschema.Schema{ForceNew: true},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE},
+		},
+		{
+			name: "ps force new unchanged",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("old"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: nil,
+		},
+		{
+			name: "ps force new changed non-nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewStringProperty("new"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE},
+		},
+		{
+			name: "ps force new changed from nil",
+			old:  resource.NewNullProperty(),
+			new:  resource.NewStringProperty("new"),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE},
+		},
+		{
+			name: "ps force new changed to nil",
+			old:  resource.NewStringProperty("old"),
+			new:  resource.NewNullProperty(),
+			eps:  &SchemaInfo{ForceNew: True()},
+			want: &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := detailedDiffer{
+				tfs: shimschema.SchemaMap{"foo": tt.etf.Shim()},
+				ps:  map[string]*SchemaInfo{"foo": tt.eps},
+			}.makeTopPropDiff(tt.old, tt.new, newPropertyPath("foo"))
+			if got == nil && tt.want == nil {
+				return
+			}
+			if got == nil || tt.want == nil {
+				t.Errorf("makeTopPropDiff() = %v, want %v", got, tt.want)
+				return
+			}
+			if got.Kind != tt.want.Kind {
+				t.Errorf("makeTopPropDiff() = %v, want %v", got.String(), tt.want.String())
+			}
+		})
+	}
+}
+
+var Added = map[string]*pulumirpc.PropertyDiff{
+	"foo": {Kind: pulumirpc.PropertyDiff_ADD},
+}
+
+var Updated = map[string]*pulumirpc.PropertyDiff{
+	"foo": {Kind: pulumirpc.PropertyDiff_UPDATE},
+}
+
+var Deleted = map[string]*pulumirpc.PropertyDiff{
+	"foo": {Kind: pulumirpc.PropertyDiff_DELETE},
+}
+
+var ComputedVal = resource.NewComputedProperty(resource.Computed{Element: resource.NewStringProperty("")})
```

Comment:
Can we move these into the tests that use them or turn them into (immutable) functions:

```go
func added() map[string]*pulumirpc.PropertyDiff{
	return map[string]*pulumirpc.PropertyDiff{
		"foo": {Kind: pulumirpc.PropertyDiff_ADD},
	}
}
```

Introducing global mutable state in tests just seems unnecessary, with a high potential for debugging later.